bpf/ut/pol_prog_test.go

```diff
@@ -846,9 +854,55 @@ var polProgramTests = []polProgramTest{
 			"setB": {"123.0.0.1/32,udp:1024"},
 		},
 	},
-
-	// TODO ICMP
-
+	//ICMP tests
+	{
+		PolicyName: "allow icmp packet with type 8",
+		Policy: [][][]*proto.Rule{{{{
+			Action: "Allow",
+			Icmp:   &proto.Rule_IcmpType{IcmpType: 8},
+		}}}},
+		AllowedPackets: []packet{
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 8, 0)},
+		DroppedPackets: []packet{
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 10, 0)},
+	},
+	{
+		PolicyName: "allow icmp packet with type 8 and code 3",
+		Policy: [][][]*proto.Rule{{{{
+			Action: "Allow",
+			Icmp:   &proto.Rule_IcmpTypeCode{&proto.IcmpTypeAndCode{Type: 8, Code: 3}},
+		}}}},
+		AllowedPackets: []packet{
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 8, 3)},
+		DroppedPackets: []packet{
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 10, 0),
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 10, 3),
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 8, 4)},
+	},
+	{
+		PolicyName: "allow icmp packet with type not equal to 8",
+		Policy: [][][]*proto.Rule{{{{
+			Action:  "Allow",
+			NotIcmp: &proto.Rule_NotIcmpType{NotIcmpType: 8},
+		}}}},
+		AllowedPackets: []packet{
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 10, 0)},
+		DroppedPackets: []packet{
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 8, 0)},
+	},
+	{
+		PolicyName: "allow icmp packet with type not equal to 8 and code not equal to 3",
+		Policy: [][][]*proto.Rule{{{{
+			Action:  "Allow",
+			NotIcmp: &proto.Rule_NotIcmpTypeCode{&proto.IcmpTypeAndCode{Type: 8, Code: 3}},
+		}}}},
+		AllowedPackets: []packet{
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 10, 0)},
+		DroppedPackets: []packet{
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 8, 3),
+			icmpPkt_with_type_code("10.0.0.1", "10.0.0.2", 8, 4),
```

Comment:
I think this packet should be allowed; I think the match should be `!((type == 8) && (code == 3))` so `code=4` should make the packet accepted